registry/storage/driver/gcs/gcs.go

```diff
@@ -234,82 +253,90 @@ func (d *driver) WriteStream(context ctx.Context, path string, offset int64, rea
 		return 0, storagedriver.InvalidOffsetError{Path: path, Offset: offset}
 	}
 
-	if offset == 0 {
-		return d.writeCompletely(context, path, 0, reader)
-	}
-
-	service, err := storageapi.New(d.client)
-	if err != nil {
-		return 0, err
-	}
-	objService := storageapi.NewObjectsService(service)
-	var obj *storageapi.Object
-	err = retry(5, func() error {
-		o, err := objService.Get(d.bucket, d.pathToKey(path)).Do()
-		obj = o
-		return err
-	})
-	//	obj, err := retry(5, objService.Get(d.bucket, d.pathToKey(path)).Do)
-	if err != nil {
-		return 0, err
-	}
-
-	// cannot append more chunks, so redo from scratch
-	if obj.ComponentCount >= 1023 {
-		return d.writeCompletely(context, path, offset, reader)
+	var s *uploadSession
+	if offset > 0 {
+		s, err = d.getUploadSession(context, path)
+		if err != nil {
+			return 0, err
+		}
 	}
-
-	// skip from reader
-	objSize := int64(obj.Size)
-	nn, err := skip(reader, objSize-offset)
-	if err != nil {
-		return nn, err
+	if s == nil {
+		s = d.newUploadSession(path)
 	}
 
-	// Size <= offset
-	partName := fmt.Sprintf("%v#part-%d#", d.pathToKey(path), obj.ComponentCount)
-	gcsContext := d.context(context)
-	wc := storage.NewWriter(gcsContext, d.bucket, partName)
-	wc.ContentType = "application/octet-stream"
+	overlap := s.length() - offset
+	if overlap > 0 {
+		totalRead, err = io.Copy(ioutil.Discard, io.LimitReader(reader, overlap))
+		if err != nil {
+			return totalRead, err
+		}
+		offset += totalRead
+	}
+
+	chunk := make([]byte, maxChunkSize)
+	more := true
+	for more {
+		copy(chunk, s.buffer)
+		start := len(s.buffer)
+		// add zeros if offset is larger than the current size of the file
+		if offset > s.length() {
+			count := len(chunk)
+			diff := offset - s.length()
+			if diff < int64(count) {
+				count = int(diff)
+			}
+			for i := 0; i < count; i++ {
+				chunk[start] = 0
+				start++
+			}
+		}
 
-	if objSize < offset {
-		err = writeZeros(wc, offset-objSize)
+		n, err := reader.Read(chunk[start:])
+		if err == io.EOF {
+			more = false
+		} else if err != nil {
+			return totalRead, err
+		}
+		remainder := (start + n) % minChunkSize
```

Comment:
Yes makes sense. 256KB is the minimal chunk size allowed by Google Cloud storage (except for last chunk).